### PR TITLE
Use HTTP over named pipes and events for callback

### PR DESCRIPTION
**Pull request by gbraad**
#### Done:
  * Communication uses HTTP over named pipe
  * Events for status received and changed, started, stopped, etc.
  * Split code into a trayicon
  * Use notifications for all messages

### CODE REVIEW

**Comment by gbraad** [comment]
I did not add the code based on the `DaemonException`
```diff
diff --git a/Communication/DaemonCommander.cs b/Communication/DaemonCommander.cs
index 6aae047..f93ae63 100755
--- a/Communication/DaemonCommander.cs
+++ b/Communication/DaemonCommander.cs
@@ -18,9 +18,28 @@ namespace CRCTray.Communication
                private static readonly string _daemonHTTPNamedPipe = "crc-http";
         private static readonly string _apiPath = "/api";

+        internal class DaemonException : Exception^M
+        {^M
+            public DaemonException(string message) : base(message)^M
+            {^M
+            }^M
+        }^M
+^M
                private static T getResultsForBasicCommand<T>(string command, int timeout = 20)
-               {
-                       return JsonSerializer.Deserialize<T>(SendBasicCommand(command, timeout));
+        {^M
+            var output = SendBasicCommand(command, timeout);^M
+            ^M
+            var options = new JsonSerializerOptions^M
+            {^M
+                IgnoreNullValues = true^M
+            };^M
+^M
+            var status = JsonSerializer.Deserialize<Result>(output, options);^M
+            if (status.Success)^M
+            {^M
+                return JsonSerializer.Deserialize<T>(output, options);^M
+            }^M
+            throw new DaemonException(status.Error);^M
                }

                public static StatusResult Status()

```

as this provides no additional benefit at this point. The status returning an 'error' has been handled very differently at the moment.

**Comment by guillaumerose** [comment]
2 things for me:
* When an action fails like start or delete, the error is not displayed anymore
* Just after install, the status is Unknown (because error 500 in the daemon - should be fixed, see https://github.com/code-ready/crc/issues/2023). A workaround is fine for now (just show the beginning of the error?).

File: Communication/DaemonCommander.cs [hunk 1]
```diff
@@ -1,58 +1,58 @@
...
+using HttpOverStream.NamedPipe;
 
 namespace CRCTray.Communication
 {
 
 	static class DaemonCommander
 	{
-		static private string socketPath = string.Format("{0}\\.crc\\crc.sock", Environment.GetFolderPath(Environment.SpecialFolder.UserProfile));
+		private static readonly string _daemonHTTPNamedPipe = "crc-http";
+        private static readonly string _apiPath = "/api";
```
Comment (gbraad) [inline Communication/DaemonCommander.cs:14]: Not sure why the indentation jumps around

image

**Comment by gbraad** [comment]
> Just after install, the status is Unknown (because error 500 in the daemon - should be fixed, see Simplify crc status: no virtual machine is equivalent to virtual machine stopped crc#2023). A workaround is fine for now (just show the beginning of the error?).

Workaround in place. In code marked as "TODO" as the issue is with the http status codes and sending a proper status code.

> When an action fails like start or delete, the error is not displayed anymore

image

They get processed according to the status returned from the daemon. If the notification is still active, the new one gets layered.

**Comment by gbraad** [comment]
> the new one gets layered.

Shortened the balloon/notification time to 3secs

**Review by anjannath — APPROVED** [review]
LGTM

**Pull request merged**